### PR TITLE
Don't write empty header frame if there are no trailers

### PR DESCRIPTION
**Pull request by valerauko**
## Motivation

Fixes https://github.com/netty/netty-incubator-codec-http3/issues/134
The empty header written fails client-side validation

## Modification

Only write header frame if there are any trailing headers

## Result

Http3FrameToHttpObjectCodec generated responses don't fail netty's validation

### CODE REVIEW

**Comment by NiteshKant** [comment]
Thanks @valerauko for the PR!

> The empty header written fails client-side validation

Can you explain more about which validation is this? Also, can you add a test that verifies the behavior?

**Comment by valerauko** [comment]
@NiteshKant When using the example client (Http3ClientExample) the response generated will throw an error,
```
io.netty.incubator.codec.http3.Http3HeadersValidationException: Not all mandatory pseudo-headers included.
```

I see some tests are failing too, I'll fix that and add a new case too if none of the fixed ones will cover this.

**Comment by NiteshKant** [comment]
> When using the example client (Http3ClientExample) the response generated will throw an error

Hmm that sounds similar to the issue (https://github.com/netty/netty-incubator-codec-http3/commit/67254b9307dbe9133923f7fa2c5bac12577c05c8) I fixed in the client example. I do not see this error on `main` anymore. Are you sure your branch was rebased when you saw this error?

**Comment by valerauko** [comment]
I don't think this is an issue in the example client code.

The server was responding with an empty trailing header frame last (without any of the pseudo-headers of course), which would fail the client's validation of header frames.

**Comment by normanmaurer** [comment]
@NiteshKant what is the status of this one ?

**Comment by valerauko** [comment]
I'll update the tests asap if the direction of this fix is okay

**Comment by NiteshKant** [comment]
@valerauko yes it does make sense to omit writing header frame if there are no trailers.

Any idea why some tests are failing after this change?

**Comment by valerauko** [comment]
I haven't looked closely yet but I think some of the tests actually _expect_ an empty header frame right now.

I'll push fixes in a day or two!

**Comment by valerauko** [comment]
@NiteshKant I've fixed the test _errors_. There are still two failures which are assertions that expect an empty header frame to be written when you write an EMPTY_LAST_CONTENT.

I don't know what's the ideal way to deal with this situation. The failing test cases explicitly expect an empty header frame to be written, but such a frame fails validation. Should I just delete the two tests that assert this? (L190 and L501 of Http3FrameToHttpObjectCodecTest).

This change also means that since no empty header frame is written, there's nothing to attach the WRITE_FIN listener to. What's ideal here?

**Comment by NiteshKant** [comment]
Thanks @valerauko 

> Should I just delete the two tests that assert this?

Yes the asserts that expect a frame to be written can be removed.

> This change also means that since no empty header frame is written, there's nothing to attach the WRITE_FIN listener to. What's ideal here?

It may be a good time to remove the usage of `WRITE_FIN` listener as it is deprecated. You can instead just call `shutdownOutput()` on the channel like:

`((QuicStreamChannel) ctx.channel()).shutdownOutput();`

**Comment by valerauko** [comment]
Thanks for the guidance. Pushed fixes.

File: src/main/java/io/netty/incubator/codec/http3/Http3FrameToHttpObjectCodec.java [hunk 1]
```diff
@@ -148,30 +148,25 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
         ChannelFuture future = null;
         if (msg instanceof HttpMessage) {
             Http3Headers headers = toHttp3Headers((HttpMessage) msg);
-             future = ctx.write(new DefaultHttp3HeadersFrame(headers));
+            future = ctx.write(new DefaultHttp3HeadersFrame(headers));
```
Comment (NiteshKant) [inline src/main/java/io/netty/incubator/codec/http3/Http3FrameToHttpObjectCodec.java]: This `future` assignment can be removed, I have removed it and pushed a commit.

**Comment by NiteshKant** [comment]
Thanks so much @valerauko

**Pull request merged**